phase2/particle
# [WIP]: Multiple Design Systems.

## WORK IN PROGRESS - DO NOT MERGE

# Multiple Design Systems

> **This description will be updated as required to define the full scope of this Pull Request to enable easier integration of multiple design systems into a single Particle instance.**

## Still @TODO
* Fix `gulpfile.js` in `compile:pl:notify:post` to somehow generate a better `index.html` document that applies to multiple design systems, not the last one generated.
* Check the [Phase2 Frontend Docs](https://phase2.gitbook.io/frontend/) for any related information, link references, best practices, etc. that need to be updated because of these changes.

## Testing Directions
Following [Quickstart B](https://github.com/phase2/particle/blob/master/README.md#quickstart-b) in [README.md](https://github.com/phase2/particle/blob/master/README.md) we will use the `feature/multi-design-system` branch in my fork at [`himerus/particle`](https://github.com/himerus/particle).

```BASH
git clone git@github.com:himerus/particle.git particle-multi-test
cd particle-multi-test
git checkout feature/multi-design-system
npm install && npm run setup
npm start
```

After the watch script is running, you can visit [http://0.0.0.0:8080/pl-default/](http://0.0.0.0:8080/pl-default/) to browse around and ensure everything is accurate.

Also test the `npm run new` command to create a few new components and ensure they work as expected.

Ensure you can make a copy of the `source/default` and `apps/pl-default` and refactoring accordingly, setup a second design system. This would include additional updates to `package.json` to ensure the npm scripts are accurate.

## This PR Accomplishes:

### Rename `apps/pl` to `apps/pl-default`.
In order that Particle implementations with multiple design systems can all be handed over to Pattern Lab, we should rename the `pl` folder to `pl-default` in the `apps` directory. This allows us to then also have `pl-design-system-two`, `pl-design-system-three` which would correspond to additional component libraries under `source`.

> We likely want to do this renaming with the `apps/drupal` and `apps/grav` apps as well.

### Update the npm scripts for watching and building PL
So that we can more efficiently add elements to our npm scripts later for additional design systems, the following changes were made:

Change:
```json
"scripts": {
    "dev:pl": "cross-env-shell NODE_ENV=development \"gulp compile:startup --config ./apps/pl/particle.app.config.js & webpack-dev-server --config ./apps/pl/webpack.config.js --hot\"",
    "build:pl": "cross-env-shell NODE_ENV=production \"gulp compile:startup --config ./apps/pl/particle.app.config.js && webpack --config ./apps/pl/webpack.config.js\"",
  },
```
To:
```json
"scripts": {
    "dev:pl": "npm run dev:pl:default",
    "dev:pl:default": "cross-env-shell NODE_ENV=development \"gulp compile:startup --config ./apps/pl-default/particle.app.config.js & webpack-dev-server --config ./apps/pl-default/webpack.config.js --hot\"",
    "build:pl": "npm run build:pl:default",
    "build:pl:default": "cross-env-shell NODE_ENV=production \"gulp compile:startup --config ./apps/pl-default/particle.app.config.js && webpack --config ./apps/pl-default/webpack.config.js\"",
  },
```
The above allows us to quickly add in new systems by doing the following:
* New line item for `dev:pl:DESIGN_SYSTEM_TWO` and `build:pl:DESIGN_SYSTEM_TWO`
* Update the default `build:pl` script to include the new design system(s) by following the example: `"build:pl": "npm run build:pl:default && npm run build:pl:DESIGN_SYSTEM_TWO",`
* Decide yourself how to handle the default watch script `dev:pl`. Should it handle only the `default` design system? Should it watch them all somehow?

### Update `source/default/namespaces.js` to use `default_` prefix on ALL namespaces.
> We need to start all new projects with namespaces that assumes the names `atoms`, `molecules`, etc. could be in conflict with something from another system, including one of our legacy `p2-theme-core` projects.

This is to avoid ANY name conflicts between design systems, regardless of the outcome of the [Components module issue](https://www.drupal.org/project/components/issues/3027672).
```JS
const sets = {
  default_protons: path.resolve(__dirname, '_patterns', '00-protons'),
  default_atoms: path.resolve(__dirname, '_patterns', '01-atoms'),
  default_molecules: path.resolve(__dirname, '_patterns', '02-molecules'),
  default_organisms: path.resolve(__dirname, '_patterns', '03-organisms'),
  default_templates: path.resolve(__dirname, '_patterns', '04-templates'),
  default_pages: path.resolve(__dirname, '_patterns', '05-pages'),
};
```
### Update ALL files to search/replace namespaced templates
Search and Replace all the following:
* `@protons` -> `@default_protons`
* `@atoms` -> `@default_atoms`
* `@molecules` -> `@default_molecules`
* `@organisms` -> `@default_organisms`
* `@templates` -> `@default_templates`
* `@pages` -> `@default_pages`
* `import 'protons` -> `import 'default_protons`
* `import 'atoms` -> `import 'default_atoms`
* `import 'organisms` -> `import 'default_organisms`
* ... etc.

### Change the default folder structure of `dist`
Rather than having the `assets` and `pl` folder live separately, we can simply use `dist/pl-default` to hold both the assets AND the rest of what PL needs. This also requires adjusting the `cleanPublic` flag  in `config.yml` to `false`. This should allow us to more closely 'mirror' the built and watched versions of PL at the same time.

If we have multiple design systems, our structure would look like:
* `dist`
  * `pl-default`
    * `assets/`
    * `index.html`
  * `pl-design-system-two`
    * `assets/`
    * `index.html`
  * `pl-design-system-three`
    * `assets/`
    * `index.html`

### Update `tools/new-component` to handle new namespacing.
* **`tools/new-component/pattern.js`:**
  * Change `import 'protons';` to `import '<%= designSystem %>_protons';`
* **`tools/new-component/demo-pattern.js`:**
  * Change `import '<%= cleanPatternType %>/<%= name %>';` to `import '<%= designSystem %>_<%= cleanPatternType %>/<%= name %>';`.
* **`tools/new-component/demo-pattern.twig`:**
  * Change `{% include '@<%= cleanPatternType %>/_<%= name %>.twig' %}` to`{% include '@<%= designSystem %>_<%= cleanPatternType %>/_<%= name %>.twig' %}';`.
